File: QytetetRuby/lib/controlador_qytetet.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'qytetet.rb'
require_relative 'jugador.rb'
require_relative 'tipo_casilla.rb'
require_relative 'vista_textual_qytetet.rb'
require_relative 'metodo_salir_carcel.rb'
module InterfazTextualQytetet
  class ControladorQytetet
    include ModeloQytetetRuby
    def initialize
      @juego = ModeloQytetetRuby::Qytetet.instance
      @vista = VistaTextualQytetet.new
    end
    def inicializacion_juego
      
      nombres = @vista.obtener_nombre_jugadores
      
      #nombres = Array.new
      #nombres<<"PepitoGrillo";
      #nombres<<"MacarenaLaPortugesa";
      #nombres<<"CamaronDeLaIsla";
      #nombres<<"NapoleonBonaparte";
      
      @juego.inicializar_juego(nombres)
      @jugador = @juego.jugador_actual
      #@vista.mostrar("Jugador Actual: "+@jugador.nombre)
      @casilla = @jugador.casilla_actual
      #@vista.mostrar("Casilla Actual: "+@casilla.numero_casilla.to_s)
      
      @vista.mostrar("#################################")
      @vista.mostrar("#####                        ####")
      @vista.mostrar("#####  Comienzo de Qytetet   ####")
      @vista.mostrar("#####                        ####")
      @vista.mostrar("#################################")
      #@vista.mostrar(@juego.to_s)
    end
    def elegir_propiedad(propiedades) # lista de propiedades a elegir
      @vista.mostrar("\tCasilla\tTitulo");
        
      listaPropiedades= Array.new
      for prop in propiedades  # crea una lista de strings con numeros y nombres de propiedades
        propString= prop.numero_casilla.to_s+' '+prop.titulo.nombre; 
        listaPropiedades<<propString
      end
      seleccion=@vista.menu_elegir_propiedad(listaPropiedades)  # elige de esa lista del menu
      propiedades.at(seleccion)
    end
    def jugador_en_carcel
      libre = false
      if (@vista.menu_salir_carcel==1)
        libre = @juego.intentar_salir_carcel(MetodoSalirCarcel::PAGANDOLIBERTAD)
      else
        libre = @juego.intentar_salir_carcel(MetodoSalirCarcel::TIRANDODADO)
      end
      if(libre)
        @vista.mostrar("El jugador "+@jugador.nombre+" ha conseguido salir de la carcel.")
      else
        @vista.mostrar("El jugador "+@jugador.nombre+" no ha conseguido salir de la carcel. Pasa el turno al siguiente jugador.");
        es_final_juego
      end
      return libre
    end
    def  gestion_inmobiliaria(encarcelado,bancarrota)
      if(!encarcelado && !bancarrota && @jugador.tengo_propiedades)
        opcion = @vista.menu_gestion_inmobiliaria
        while (opcion!=0 && @jugador.tengo_propiedades)
          
          c = elegir_propiedad(@jugador.obtener_propiedades)
          case opcion
          when 1
            if(@juego.edificar_casa(c))
              @vista.mostrar("Se ha edificado una casa en la casilla "+c.numero_casilla.to_s);
              @vista.mostrar("El saldo del jugador "+@jugador.nombre+" desciente a "+@jugador.saldo.to_s);
            else
              @vista.mostrar("No se ha edificado una casa en la casilla "+c.numero_casilla.to_s);
            end
          when 2
            if(@juego.edificar_hotel(c))
              @vista.mostrar("Se ha edificado un hotel en la casilla "+c.numero_casilla.to_s);
              @vista.mostrar("El saldo del jugador "+@jugador.nombre+" desciente a "+@jugador.saldo.to_s);
            else
              @vista.mostrar("No se ha edificado un hotel en la casilla "+c.numero_casilla.to_s);
            end
          when 3
            if(@juego.vender_propiedad(c))
              @vista.mostrar("Se ha vendido la casilla "+c.numero_casilla.to_s);
              @vista.mostrar("El saldo del jugador "+@jugador.nombre+" aumenta a "+@jugador.saldo.to_s);
            else
              @vista.mostrar("No se ha podido vender la casilla "+c.numero_casilla.to_s);
            end
          when 4
            if(@juego.hipotecar_propiedad(c))
              @vista.mostrar("Se ha hipotecado la casilla "+c.numero_casilla.to_s);
              @vista.mostrar("El saldo del jugador "+@jugador.nombre+" aumenta a "+@jugador.saldo.to_s);
            else
              @vista.mostrar("No se ha podido vender la casilla "+c.numero_casilla.to_s);
            end
          when 5
            if(@juego.cancelar_hipoteca(c))
              @vista.mostrar("Se ha cancelado la hipoteca de la casilla "+c.numero_casilla.to_s);
              @vista.mostrar("El saldo del jugador "+@jugador.nombre+" desciende a "+@jugador.saldo.to_s);
            else
              @vista.mostrar("No se ha podido cancelar la hipoteca de la casilla "+c.numero_casilla.to_s);
            end
          end
          if(@jugador.tengo_propiedades)
            opcion = @vista.menu_gestion_inmobiliaria
          end
        end
      end
    end
    def  es_final_juego
      if (@jugador.saldo>0)
        @vista.mostrar("Fin de turno. Pulsa una tecla cuando estes preparado.")
        gets.chomp
        @juego.siguiente_jugador
        @jugador = @juego.jugador_actual
        desarrollo_juego
      else
        @vista.mostrar("El jugador "+@jugador.to_s+" ha caido en bancarrota.");
        rank = @juego.obtener_ranking
        @vista.mostrar("Jugador: \t Saldo: ");
        rank.each{
          |r|
          @vista.mostrar(r.to_s)
        }
      end
    end
    def juego_tipo_sorpresa
      no_tiene_propietario = @juego.aplicar_sorpresa
      if (@jugador.encarcelado)
        @vista.mostrar("El jugador " + @jugador.to_s + " ha sido encarcelado. Pulsa para continuar.")
        gets.chomp
        es_final_juego
      end
      if (@jugador.saldo<=0)
        @vista.mostrar("El jugador " + @jugador.to_s + " ha entrador en bancarrota. Pulsa para continuar.")
        gets.chomp
        es_final_juego
      end
      if(@jugador.casilla_actual != @casilla)
        @vista.mostrar("El jugador " + @jugador.nombre + " ha cambiado de posicion, ahora esta en la casilla " + @jugador.casilla_actual.numero_casilla.to_s)
        @casilla = @jugador.casilla_actual
        if(@casilla.tipo == TipoCasilla::CALLE)
          if(!@casilla.titulo.tengo_propietario)
            if(@vista.elegir_quiero_comprar)
              @juego.comprar_titulo_propiedad
            end
          else
            @vista.mostrar("La casilla " + @casilla.to_s + " tiene propietario, el jugador " + @jugador.nombre + " tiene que pagar "+@casilla.cobrar_alquiler)
          end
        end
      end
      return no_tiene_propietario
    end
    def jugador_libre
      no_tiene_propietario = @juego.jugar
      @casilla = @jugador.casilla_actual
      @vista.mostrar("Pulsa una tecla para avanzar a la casilla: "+@casilla.numero_casilla.to_s)
      gets.chomp
      bancarrota = @jugador.saldo<0
      if(!bancarrota)
        encarcelado = @jugador.encarcelado
        if(!encarcelado)
          if(@casilla.tipo == TipoCasilla::SORPRESA)
            @vista.mostrar("El jugador "+@jugador.to_s+" ha caido en una casilla tipo sorpresa.\n");
            @vista.mostrar("Ha tocado la sopresa: "+@juego.carta_actual.to_s);
            juego_tipo_sorpresa
          elsif(@casilla.tipo == TipoCasilla::CALLE)
            no_tiene_propietario = !@casilla.titulo.tengo_propietario
            if (no_tiene_propietario)
              @vista.mostrar("La calle "<<@casilla.to_s<<" se puede comprar. Elige una opción.")
              if(@vista.elegir_quiero_comprar)
                @juego.comprar_titulo_propiedad
                @vista.mostrar("El jugador "+@jugador.to_s+" ha comprado la casilla "+@casilla.numero_casilla.to_s)
                @vista.mostrar(@jugador.nombre+" se ha quedado con un saldo total de "+@jugador.saldo.to_s)
              else
                @vista.mostrar("El jugador "+@jugador.to_s+" no ha comprado la casilla "+@casilla.numero_casilla.to_s)
              end
            else
              if(!@casilla.esta_hipotecada)
                @vista.mostrar("La casilla "+@casilla.numero_casilla.to_s+" tiene propietario, el jugador "+@jugador.nombre+" tiene que pagar "+@casilla.cobrar_alquiler.to_s)
                @vista.mostrar(@jugador.nombre+" se ha quedado con un saldo total de "+@jugador.saldo.to_s)
              end
            end
          end
        end
          gestion_inmobiliaria(encarcelado,bancarrota)
          es_final_juego
      end
    end
    def desarrollo_juego
      libre = true
      @jugador = @juego.jugador_actual
      @casilla = @jugador.casilla_actual
      
      @vista.mostrar("Turno del jugador: "+@jugador.nombre);
      @vista.mostrar("Esta en la casilla: "+@casilla.numero_casilla.to_s);
      @vista.mostrar("Es una casilla de tipo: "+@casilla.tipo.to_s);
       
      if (@jugador.encarcelado)
        libre = jugador_en_carcel;
      end
      if (libre)
        jugador_libre
      end
    end
    def self.main
      controlador = ControladorQytetet.new
      controlador.inicializacion_juego
      controlador.desarrollo_juego
    end
    
    
  end
  
  ControladorQytetet.main
end

File: QytetetJava/src/InterfazTextualQytetet/ControladorQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package InterfazTextualQytetet;

import java.util.ArrayList;
import java.util.HashMap;
import modeloqytetet.*;

/**
 *
 * @author Migue
 */
public class ControladorQytetet {

    static Qytetet juego;
    static Jugador jugador;
    static Casilla casilla;
    static VistaTextualQytetet vista = new VistaTextualQytetet();

    public static void inicializacionJuego() {
        juego = Qytetet.getInstance();
        //ArrayList<String> nombres = vista.obtenerNombreJugadores();
        ArrayList<String> nombres = new ArrayList();

        nombres.add("PepitoGrillo");
        nombres.add("MacarenaLaPortugesa");
        nombres.add("CamaronDeLaIsla");
        nombres.add("NapoleonBonaparte");

        juego.inicializarJuego(nombres);

        jugador = juego.getJugadorActual();
        casilla = jugador.getCasillaActual();

        vista.mostrar(juego.toString());
    }

    public static Casilla elegirPropiedad(ArrayList<Casilla> propiedades) {
        //este metodo toma una lista de propiedades y genera una lista de strings, con el numero y nombre de las propiedades
        //luego llama a la vista para que el usuario pueda elegir.
        vista.mostrar("\tCasilla\tTitulo");
        int seleccion;
        ArrayList<String> listaPropiedades = new ArrayList();
        for (Casilla casill : propiedades) {
            listaPropiedades.add("\t" + casill.getNumeroCasilla() + "\t" + ((Calle)casill).getTituloPropiedad().getNombre());
        }
        seleccion = vista.menuElegirPropiedad(listaPropiedades);
        return propiedades.get(seleccion);
    }

    private static boolean JugadorEnCarcel() {
        boolean libre = false;
        if (vista.menuSalirCarcel() == 1) {
            libre = juego.intentarSalirCarcel(MetodoSalirCarcel.PAGANDOLIBERTAD);
        } else {
            libre = juego.intentarSalirCarcel(MetodoSalirCarcel.TIRANDODADO);
        }

        if (libre) {
            vista.mostrar("El jugador " + jugador.getNombre() + " ha conseguido salir de la carcel.");
        } else {
            vista.mostrar("El jugador " + jugador.getNombre() + " no ha conseguido salir de la carcel. Pasa el turno al siguiente jugador.");
        }
        return libre;
    }

    private static boolean juegoTipoSorpresa() {
        boolean noTienePropietario = juego.aplicarSorpresa();
        if (jugador.getEncarcelado()) {
            vista.mostrar("El jugador " + jugador.getNombre() + " ha sido encarcelado. Pulsa para continuar.");
            vista.in.nextLine();
            esFinalDeJuego();
        }
        if (jugador.getSaldo() <= 0) {
            vista.mostrar("El jugador " + jugador.getNombre() + " ha entrador en bancarrota. Pulsa para continuar.");
            vista.in.nextLine();
            esFinalDeJuego();
        }
        if (jugador.getCasillaActual() != casilla) {
            vista.mostrar("El jugador " + jugador.getNombre() + " ha cambiado de posicion, ahora esta en la casilla " + jugador.getCasillaActual());
            casilla = jugador.getCasillaActual();
            if (casilla.getTipo() == TipoCasilla.CALLE) {
                if (!((Calle)casilla).getTituloPropiedad().tengoPropietario()) {
                    if (vista.elegirQuieroComprar()) {
                        juego.comprarTituloPropiedad();
                    }
                } else {
                    vista.mostrar("La casilla " + casilla.getNumeroCasilla() + " tiene propietario, el jugador " + jugador.getNombre() + " tiene que pagar.");
                    vista.mostrar(jugador.getNombre() + " se ha quedado con un saldo total de " + jugador.getSaldo());
                }
            }
        }

        return noTienePropietario;
    }

    private static void GestionInmobiliaria(boolean encarcelado, boolean bancarrota) {
        if (!encarcelado && !bancarrota && jugador.tengoPropiedades()) {
            int opcion = vista.menuGestionInmobiliaria();
            while (opcion != 0){
                Calle c;
                c = (Calle)elegirPropiedad(jugador.obtenerPropiedades());
                switch (opcion) {
                    case 1:
                        if (juego.edificarCasa(c)) {
                            vista.mostrar("Se ha edificado una casa en la casilla " + c.getNumeroCasilla());
                            vista.mostrar("El saldo del jugador " + jugador.getNombre() + " desciente a " + jugador.getSaldo());
                        } else {
                            vista.mostrar("No se ha edificado una casa en la casilla " + c.getNumeroCasilla());
                        }
                        break;
                    case 2:
                        if (juego.edificarHotel(c)) {
                            vista.mostrar("Se ha edificado un hotel en la casilla " + c.getNumeroCasilla());
                            vista.mostrar("El saldo del jugador " + jugador.getNombre() + " desciente a " + jugador.getSaldo());
                        } else {
                            vista.mostrar("No se ha edificado un hotel en la casilla " + c.getNumeroCasilla());
                        }
                        break;
                    case 3:
                        if (juego.venderPropiedad(c)) {
                            vista.mostrar("Se ha vendido la casilla " + c.getNumeroCasilla());
                            vista.mostrar("El saldo del jugador " + jugador.getNombre() + " aumenta a " + jugador.getSaldo());
                        } else {
                            vista.mostrar("No se ha podido vender la casilla " + c.getNumeroCasilla());
                        }
                        break;
                    case 4:
                        if (juego.hipotecarPropiedad(c)) {
                            vista.mostrar("Se ha hipotecado la casilla " + c.getNumeroCasilla());
                            vista.mostrar("El saldo del jugador " + jugador.getNombre() + " aumenta a " + jugador.getSaldo());
                        } else {
                            vista.mostrar("No se ha podido vender la casilla " + c.getNumeroCasilla());
                        }
                        break;
                    case 5:
                        if (juego.cancelarHipoteca(c)) {
                            vista.mostrar("Se ha cancelado la hipoteca de la casilla " + c.getNumeroCasilla());
                            vista.mostrar("El saldo del jugador " + jugador.getNombre() + " desciende a " + jugador.getSaldo());
                        } else {
                            vista.mostrar("No se ha podido cancelar la hipoteca de la casilla " + c.getNumeroCasilla());
                        }
                        break;
                }
                opcion = vista.menuGestionInmobiliaria();
            }
        }
    }

    private static void esFinalDeJuego() {
        if (jugador.getSaldo() > 0 /*&& prueba*/) {
            vista.mostrar("Fin de turno. Pulsa una tecla cuando estes preparado.");
            vista.in.nextLine();
            juego.siguienteJugador();
            jugador = juego.getJugadorActual();
            desarrolloJuego();
        } else {
            vista.mostrar("El jugador " + jugador.getNombre() + " ha caido en bancarrota.");
            HashMap rank = juego.obtenerRanking();
            rank.forEach((k, v) -> vista.mostrar("Jugador: " + k + ": Saldo: " + v));
        }

    }

    private static void JugadorLibre() {
        boolean noTienePropietario = juego.jugar();
        casilla = jugador.getCasillaActual();
        vista.mostrar("Pulsa una tecla para avanzar a la casilla: " + casilla.getNumeroCasilla());
        vista.in.nextLine();
        boolean bancarrota = jugador.getSaldo() < 0;

        if (!bancarrota) {
            boolean encarcelado = jugador.getEncarcelado();
            if (!encarcelado) {
                TipoCasilla tipo = casilla.getTipo();
                if (tipo == TipoCasilla.SORPRESA) {
                    vista.mostrar("El jugador " + jugador.getNombre() + " ha caido en una casilla tipo sorpresa.\n");
                    vista.mostrar("Ha tocado la sopresa: " + juego.getCartaActual().getTexto());
                    //noTienePropietario = 
                    juegoTipoSorpresa();

                } else if (tipo == TipoCasilla.CALLE) {
                    noTienePropietario = !((Calle)casilla).getTituloPropiedad().tengoPropietario();
                    if (noTienePropietario) {
                        vista.mostrar("La casilla " + casilla.getNumeroCasilla() + " se puede comprar. Elige una opción.");
                        if (vista.elegirQuieroComprar()) {
                            juego.comprarTituloPropiedad();
                            vista.mostrar("El jugador " + jugador.getNombre() + " ha comprado la casilla " + casilla.getNumeroCasilla());
                            vista.mostrar(jugador.getNombre() + " se ha quedado con un saldo total de " + jugador.getSaldo());
                        } else {
                            vista.mostrar("El jugador " + jugador.getNombre() + " no ha comprado la casilla " + casilla.getNumeroCasilla());
                        }
                    } else if (!((Calle)casilla).estaHipotecada()) {
                        vista.mostrar("La casilla " + casilla.getNumeroCasilla() + " tiene propietario, el jugador " + jugador.getNombre() + " tiene que pagar.");
                        vista.mostrar(jugador.getNombre() + " se ha quedado con un saldo total de " + jugador.getSaldo());
                    }
                }
                //Gestion Inmobiliaria
                GestionInmobiliaria(encarcelado, bancarrota);
                esFinalDeJuego();
            }
        }
    }

    public static void desarrolloJuego() {
        boolean libre = true;

        jugador = juego.getJugadorActual();
        casilla = jugador.getCasillaActual();

        vista.mostrar("Turno del jugador: " + jugador.getNombre());
        vista.mostrar("Esta en la casilla: " + casilla.getNumeroCasilla());
        vista.mostrar("Es una casilla de tipo: " + casilla.getTipo());

        //jugador.prueba();
        if (jugador.getEncarcelado()) {
            libre = JugadorEnCarcel();
        }

        if (libre) {
            JugadorLibre();
        }
    }

    public static void main(String[] args) {
        inicializacionJuego();
        desarrolloJuego();
    }
}
